File: trees/938_RangeSumBST.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        values = []
        def traverse(node: Optional[TreeNode]):
            if node:
                if node.val <= high and node.val >= low:
                    values.append(node.val)
                    traverse(node.left)
                    traverse(node.right)
                elif node.val < low:
                    traverse(node.right)
                elif node.val > high:
                    traverse(node.left)
        traverse(root)
        return sum(values)

class IterSolution():
    """Doing too much recursion need to maybe try iter for trees too"""
    def rangeSumBST(self, root, L, R):
        ans = 0
        stack = [root]
        while stack:
            node = stack.pop()
            if node:
                if L <= node.val <= R:
                    ans += node.val
                if L < node.val:
                    stack.append(node.left)
                if node.val < R:
                    stack.append(node.right)
        return ans

File: trees/938_RangeSumBST.cc
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    int sum = 0;
public:
    int rangeSumBST(TreeNode* root, int low, int high) {
        this->traverse(root, low, high);
        return this->sum;
    }

    void traverse(TreeNode* node, int low, int high){
        if(node == nullptr) return;
        if(node->val <= high && node->val >= low){
            sum += node->val;
        }
        traverse(node->left, low, high);
        traverse(node->right, low, high);
    }
};